My solution to the exercise

shaharsat opened the pull request.

shaharsat committed:
The solution

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func main() {
+	server, err := setupServer()
```
Let's leave only the main function in the main file.
I recommend creating routes.go and setup.go files.
Think where to put them.

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+func setupServer() (*handlers.Server, error) {
+	elasticUrl := os.Getenv("ELASTIC_URL")
```
Don't use env variables here, we don't need it.
Const those values.

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+	})
+
+	server := handlers.Server{
+		ElasticClient: elasticClient,
```
Good idea to initiate the clients once on setup.
1. We don't use Println for errors. In this case, where ES or Redis failed to connect it's best to panic.
2. I don't like how you combined both clients into one struct as their functionality should be separated from one another in this case. 
Also, avoid generic names like Server.

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+func setupRoutes(server *handlers.Server) *gin.Engine {
+	routes := gin.Default()
+
+	cachedRoutes := routes.Group("/")
```
Redundant /

YuvalSteiy replied:
Let's use curly braces and indentation for the group - 
```
cachedRoutes := routes.Group("/")
{
    cachedRoutes.PUT(...)
    cachedRoutes.GET(...)
    etc...
}
```

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+	cachedRoutes.DELETE("/book", server.DeleteById)
+	cachedRoutes.GET("/search", server.Search)
+	cachedRoutes.GET("/stores", server.Store)
+	routes.GET("/activity", server.Activity)
```
Put this first before the cachedRoutes group

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+	cachedRoutes := routes.Group("/")
+	cachedRoutes.Use(server.CacheUserRequest)
+
+	cachedRoutes.PUT("/book", server.CreateBook)
```
Don't make the route handler function a struct method.
This is bad practice for bigger servers who don't just have one "service" entity.
Inside the handler function you can instantiate structs and use their methods.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+}
+
+func (server *Server) CreateBook(c *gin.Context) {
+	title := c.PostForm("title")
```
The request body should be JSON

YuvalSteiy replied:
You defined the JSON field names for the book struct.
Maybe you can someway unmarshal them more easily.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		return
+	}
+
+	priceF, err := strconv.ParseFloat(price, 64)
```
We don't use such names in go. Adding "F" for float etc...
Only in ES.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		return
+	}
+
+	id := uuid.New().String()
```
redundant. ES will give the document an id. You don't have to take care of this.

YuvalSteiy commented on models/book.go:
```diff
@@ -0,0 +1,39 @@
+)
+
+type Book struct {
+	Id             string     `json:"index._id" binding:"required"`
```
What is this nested field? why do you need two id fields for each document?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	"time"
+)
+
+type Server struct {
```
Think about how to create separation between ES and Redis functionality.
Make them both autonomous, so that the Redis cache for example can cache requests without "knowing" about if they use ES or what they do.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+
+func (server *Server) CreateBook(c *gin.Context) {
+	title := c.PostForm("title")
+	authorName := c.PostForm("authorName")
```
url fields should be named with under_score, Go names with camelCase.
```suggestion
	authorName := c.PostForm("author_name")
```

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+}
+
+func (server *Server) UpdateBookTitleById(c *gin.Context) {
+	id, idOk := c.GetPostForm("id")
```
same, don't use PostForm. Request body should be JSON.

YuvalSteiy replied:
In all routes where we perform an operation by id on a book (GET, UPDATE, DELETE) and we must get the id, I would add it as a query param. 
`/book/:id`

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		return
+	}
+
+	scriptFormat := "ctx._source.%s = \"%v\"" // Is this vulnerable to ElasticSearch Injection?
```
Perhaps when used like that...
Scripts give us much less efficient performance with ES and we don't use them unless we must.
Do you have an option here?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		Route:  c.Request.URL.Path,
+	}
+
+	username, ok := c.GetQuery("username")
```
Let's add an option to query without a username.
Of course it will not be cached.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	doc, err := server.ElasticClient.
+		Search().
+		Index("books").
+		Query(elastic.NewTermQuery("index._id", id)).
```
what is `index._id`?
why is it empty in the response?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+func (server *Server) GetBookById(c *gin.Context) {
+	id, idOk := c.GetQuery("id")
+
+	missingFields := make([]string, 0)
```
You just have one field here, why do all this?
Also see my comment on passing id as query parameter.

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+	cachedRoutes.PUT("/book", server.CreateBook)
+	cachedRoutes.POST("/book", server.UpdateBookTitleById)
+	cachedRoutes.GET("/book", server.GetBookById)
+	cachedRoutes.DELETE("/book", server.DeleteById)
```
Try and use consistent names.
If before you had `GetBookById` then this should be `DeleteBookById`

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	doc, err := server.ElasticClient.
+		Delete().
+		Index("books").
+		Id(id).
```
This route works with the real id but the `getBook` route works with index._id?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	c.JSON(http.StatusOK, gin.H{"status": "deleted", "id": doc.Id})
+}
+
+func (server *Server) Search(c *gin.Context) {
```
When I search for a book with a given title, I want to get only books with the exact same title in the response.
When I search using title and price range, I want to get only books with the exact same title that are in this price range.
How would you implement this?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	title, titleOk := c.GetQuery("title")
+	authorName, authorNameOk := c.GetQuery("authorName")
+	price, priceOk := c.GetQuery("price")
+	ebookAvailable, ebookAvailableOk := c.GetQuery("ebookAvailable")
```
This route should only handle the fields `title`, `author_name` and  `price_range` according to the assignment.

YuvalSteiy commented on main.go:
```diff
@@ -0,0 +1,61 @@
+	cachedRoutes.GET("/book", server.GetBookById)
+	cachedRoutes.DELETE("/book", server.DeleteById)
+	cachedRoutes.GET("/search", server.Search)
+	cachedRoutes.GET("/stores", server.Store)
```
Should be called `store`

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		PublishDate:    models.CustomTime(publishDateT),
+	}
+
+	_, err = server.ElasticClient.Index().Index("books").BodyJson(book).Do(c)
```
Const the index name

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+func (server *Server) Store(c *gin.Context) {
+	query := server.ElasticClient.Search().
+		Index("books").
+		Query(elastic.NewMatchAllQuery())
```
Why do you need this?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		return
+	}
+
+	var searchResult models.SearchResult
```
Don't need this struct.
Use `results.Aggregations.Cardinality(...)`

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	}
+
+	obj, _ := json.Marshal(operation)
+	// Not failing a request if there's a problem caching it
```
But if there is an error you don't want to try pushing obj to redis.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+func (server *Server) Activity(c *gin.Context) {
+	username := c.Query("username")
+
+	get, err := server.RedisClient.LRange(username, 0, 2).Result()
```
get is not a good name

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	var ops [3]models.UserRequest
+
+	for idx, val := range get {
+		err := json.Unmarshal([]byte(val), &ops[idx])
```
Append each time.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+		return
+	}
+
+	var ops [3]models.UserRequest
```
Lets init this as an empty slice
```suggestion
	ops := make(models.UserRequest, 0)
```

YuvalSteiy commented on models/book.go:
```diff
@@ -0,0 +1,39 @@
+
+type CustomTime time.Time
+
+const ctLayout = "2006-01-02"
```
You use this "2006-01-02" in another place as well.
Make it public and use the const there too.
ctLayout is a bad name, what about `timeFormat`?

YuvalSteiy commented on models/search_results.go:
```diff
@@ -0,0 +1,10 @@
+package models
```
You don't need this file.

YuvalSteiy commented on models/book.go:
```diff
@@ -0,0 +1,39 @@
+	AuthorName     string     `json:"author_name"  binding:"required"`
+	Price          float64    `json:"price"  binding:"required"`
+	EbookAvailable bool       `json:"ebook_available"  binding:"required"`
+	PublishDate    CustomTime `json:"publish_date" binding:"required"`
```
CustomTime is not a good name.

YuvalSteiy commented on models/book.go:
```diff
@@ -0,0 +1,39 @@
+const ctLayout = "2006-01-02"
+
+// UnmarshalJSON Parses the json string in the custom format
+func (ct *CustomTime) UnmarshalJSON(b []byte) (err error) {
```
Why do you need this? Can't you just use time?

shaharsat replied:
The assignment requests use "2006-01-02" format which requires special marshal and unmarshal functions to work

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,369 @@
+	ebookAvailable := c.PostForm("ebookAvailable")
+	publishDate := c.PostForm("publishDate")
+
+	validationError := server.validateCreateBook(title, authorName, price, ebookAvailable, publishDate)
```
Let's make all fields except "title" optional.

YuvalSteiy reviewed (changes requested):
Overall nice work!
Make these changes and we will have another iteration.

shaharsat committed:
Converted the Server methods to Functions

shaharsat committed:
Removed the searchResults struct and replaced it by constructing the response object manually

shaharsat committed:
Replaced UUID with ES's ID

shaharsat committed:
Replaced the script update with a regular update

shaharsat committed:
- Fixed a bug in Search where only one query parameter was used for search
- Search accepts JSON instead of Post Parameters

shaharsat committed:
- Renamed DeleteById to DeleteBooksById
- Renamed Search to SearchBooks

This was done to maintain a consistent naming scheme

shaharsat committed:
typo in query variable: author_ame -> author_name

shaharsat committed:
Made every parameter except 'title' optional

shaharsat committed:
removed println

YuvalSteiy commented on cmd/routes.go:
```diff
@@ -0,0 +1,25 @@
+package cmd
```
We only put main.go files under cmd package if there is one.

YuvalSteiy commented on cmd/setup.go:
```diff
@@ -0,0 +1 @@
+package cmd
```
Empty file?

YuvalSteiy commented on models/redis_connector.go:
```diff
@@ -0,0 +1,16 @@
+package models
```
redis_connector does not belong in models.

YuvalSteiy commented on models/elastic_connector.go:
```diff
@@ -0,0 +1,20 @@
+package models
```
elastic_connector does not belong in models.

YuvalSteiy commented on models/elastic_connector.go:
```diff
@@ -0,0 +1,20 @@
+func SetupElasticSearch() {
+	elasticUrl := os.Getenv("ELASTIC_URL")
+
+	elasticClient, err := elastic.NewClient(elastic.SetURL(elasticUrl))
```
```suggestion
	ElasticClient, err := elastic.NewClient(elastic.SetURL(elasticUrl))
```

YuvalSteiy commented on models/operation.go:
```diff
@@ -0,0 +1,6 @@
+package models
```
rename file to user_request

YuvalSteiy commented on models/book.go:
```diff
@@ -0,0 +1,38 @@
+type Date time.Time
+
+// UnmarshalJSON Parses the json string in the custom format
+func (ct *Date) UnmarshalJSON(b []byte) (err error) {
```
Why do you need this?

shaharsat replied:
According to the pattern in the spec, the date format is 2006-02-01. Thus both ES and the Book class must handle this format. For JSON Marshaling to work with this date format a custom date type is required.

YuvalSteiy commented on cmd/routes.go:
```diff
@@ -0,0 +1,25 @@
+
+	cachedRoutes := routes.Group("/")
+	{
+		cachedRoutes.Use(handlers.CacheUserRequest)
```
Maybe create some separation here since this is not a route.. WDYT?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+
+	c.JSON(http.StatusOK, gin.H{
+		"status": "created",
+		"id":     doc.Id,
```
Books in your index still have the "index._id" field. Please remove it from all.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+
+func CreateBook(c *gin.Context) {
+	var book models.Book
+	if err := c.ShouldBindJSON(&book); err != nil {
```
🚀

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+func SearchBooks(c *gin.Context) {
+	title := c.Query("title")
+	authorName := c.Query("author_name")
+	minPrice, minPriceOk := c.GetQuery("min_price")
```
You don't need the ok values since you check that it is not an empty string anyway. Use `c.DefaultQuery()` instead.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+	maxPrice, maxPriceOk := c.GetQuery("max_price")
+
+	if title == "" && authorName == "" && minPrice == "" && maxPrice == "" {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": "at least one query parameter is required"})
```
Instead of saying that a query parameter is required, I would say that some field value is required in order to search.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+
+	index := models.ElasticClient.Search().Index(INDEX_NAME).Pretty(false).Size(10000)
+
+	boolQuery := elastic.NewBoolQuery()
```
It doesn't work well. 
When I give a specific book title I want in return only the one book with that exact title, not many books ranked by textual relevance. What is the difference in the ES query?

shaharsat replied:
`http://localhost:8080/search?title=do`

I am not sure I understand this comment. This query returns all books with a title that contains the word "do". And in general this words for any substring of complete words. It doesn't work for general substring. This could be fixed using a wildcard query of a different analyzer for the title field.

YuvalSteiy replied:
This should return only books whose title is "do", not contains do.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+	shouldIncludePriceRangeQuery := false
+	if minPriceOk {
+		price, err := strconv.ParseFloat(minPrice, 64)
+		if err == nil {
```
This order seems less natural. Let's check `if err!=nil`, that way we don't need the `else` clause right ?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+	if maxPriceOk {
+		price, err := strconv.ParseFloat(maxPrice, 64)
+		if err == nil {
+			index.Query(priceRangeQuery.Lte(price))
```
If you are adding the price range query to be a part of the bool query, then what is this?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+		return
+	}
+
+	books := make([]models.Book, len(result.Hits.Hits))
```
You don't have to check `len()`. ES returns this information under `result.Hits.TotalHits.Value`.

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+		return
+	}
+
+	ops := make([]models.UserRequest, 0)
```
I don't like the use of the name ops here so much, since we are already calling it user requests in every other place.
WDYT?
Maybe in line 204 we can call it `userRequestsRaw`? or do you have another idea?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func CacheUserRequest(c *gin.Context) {
+	operation := models.UserRequest{
```
```suggestion
	request := models.UserRequest{
```

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+	}
+
+	obj, _ := json.Marshal(operation)
+	// Not failing a request if there's a problem caching it
```
You didn't reply to my message here. If there is an error marshaling we don't want to try and push it into Redis right?
Will obj be nil in this case? if so LPUSH will make a nil pointer dereference.

YuvalSteiy replied:
Also, I don't think obj is such a good name here since we are encoding (marshaling) an object into a JSON. This is an encoded JSON of an object. I would call an unmarshalled struct an object, no?

YuvalSteiy commented on handlers/handlers.go:
```diff
@@ -0,0 +1,244 @@
+		return
+	}
+
+	_, err := models.ElasticClient.
```
What if we will suddenly want to implement the book store search logic not using ElasticSearch?
We will have to rewrite this whole file.
Let's make it more generic - we can use an interface with all these methods and have handlers.go file not know the underlying implementation.
We can do this for the Redis too but in a separate interface, like we said it should be a cache that is autonomous to the logic of handling the request.

shaharsat committed:
Put the elastic specific code in an struct by implementing an interface

shaharsat committed:
Made the code clearer and removed unused variables

shaharsat committed:
Created an interface and struct for cache

shaharsat committed:
Renamed files and structs to make the code more readable

shaharsat committed:
Handling Redis commands

shaharsat committed:
Changed internal->service and replaced a title Match query with a Term Query

shaharsat committed:
Changed project structure according to convention

shaharsat committed:
Fixed the term query

YuvalSteiy commented on db/library_database_manager.go:
```diff
@@ -0,0 +1,14 @@
+
+type Id string
+
+type LibraryDatabaseManager interface {
```
WDYT about just `LibraryManager` ?

YuvalSteiy commented on db/library_elastic_database.go:
```diff
@@ -0,0 +1,143 @@
+	"strconv"
+)
+
+type ElasticLibraryDatabase struct {
```
And for this `ElasticLibraryManager`

YuvalSteiy commented on cache/library_cache_manager.go:
```diff
@@ -0,0 +1,6 @@
+package cache
+
+type LibraryCacheManager interface {
```
WDYT about `RequestCacher`

YuvalSteiy commented on cache/library_redis_cache.go:
```diff
@@ -0,0 +1,31 @@
+
+import "gin/config"
+
+type LibraryRedisCache struct {
```
and `RedisRequestCacher`

YuvalSteiy commented on service/handlers.go:
```diff
@@ -0,0 +1,172 @@
+const INDEX_NAME = "books"
+const MAX_NUMBER_CACHED = 3
+
+var ElasticLibrary = db.CreateElasticLibrary(INDEX_NAME)
```
If you have a global ElasticLibrary instance it should be inside the `db` folder. 
Pass the ES client to the ElasticLibrary as a dependency injection instead of having them separate.
I prefer not to use this kind of initialization.
WDYT about Initiating it in the main before the routes.

YuvalSteiy commented on service/handlers.go:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	request, err := json.Marshal(userRequest)
+	if err != nil {
```
Maybe you mean NOT to write if err != nil.
It doesn't work now.

YuvalSteiy commented on db/library_database_manager.go:
```diff
@@ -0,0 +1,14 @@
+
+import "gin/models"
+
+type Id string
```
Let's just use String

YuvalSteiy commented on service/handlers.go:
```diff
@@ -0,0 +1,172 @@
+	maxPrice := c.Query("max_price")
+
+	if title == "" && authorName == "" && minPrice == "" && maxPrice == "" {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": "at least one query parameter is required for search"})
```
Still don't love this error message.
What about - "No field to search by found" or something like that

YuvalSteiy commented on db/library_elastic_database.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func (library *ElasticLibraryDatabase) Search(title, authorName, minPrice, maxPrice string) ([]*models.Book, error) {
+	index := config.ElasticClient.Search().Index(library.IndexName).Pretty(false).Size(10000)
```
Maybe start with building the bool query completely.  After that write the entire query request to ES together like in other places.

YuvalSteiy commented on db/library_elastic_database.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func (library *ElasticLibraryDatabase) Store() (map[string]interface{}, error) {
+	query := config.ElasticClient.Search().
```
Same like search. Get everything you need and then write the query request.

YuvalSteiy commented on db/library_elastic_database.go:
```diff
@@ -0,0 +1,143 @@
+	numberOfBooks, _ := results.Aggregations.Cardinality("number_of_books")
+	numberOfAuthors, _ := results.Aggregations.Cardinality("number_of_authors")
+
+	return map[string]interface{}{
```
Lets return just the fields here and format the data in the handler. Also, why here not use gin.H{} like other places?

YuvalSteiy commented on db/library_elastic_database.go:
```diff
@@ -0,0 +1,143 @@
+	}, nil
+}
+
+func (library *ElasticLibraryDatabase) Create(book *models.Book) (Id, error) {
```
IMO it's convenient and organized if the order of the functions here is like in the handlers.go file and in the routes definition. If you want you can change.

YuvalSteiy commented on service/handlers.go:
```diff
@@ -0,0 +1,172 @@
+		c.AbortWithStatusJSON(t.Status, gin.H{"message": t.Error()})
+		return
+	case error:
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
```
If it's not an ES error, by the code it must be an unmarshalling error.
How is this a bad request? 
If you mean the case where zero docs were found with this id, I would consider using `errors.New()` inside `ElasticLibrary.GetById()` and not rely on the unmarshal error for this.

YuvalSteiy commented on service/handlers.go:
```diff
@@ -0,0 +1,172 @@
+		return
+	}
+
+	c.JSON(
```
```suggestion
	c.JSON(http.StatusOk, store)
```

YuvalSteiy commented on service/handlers.go:
```diff
@@ -0,0 +1,172 @@
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{"status": "deleted"})
```
Let's const the word "status" since we use it a lot. Maybe as `statusKey`. You can also create consts for `statusDeleted` etc...

shaharsat committed:
Replaced Id with string

shaharsat committed:
Replaced string literals with consts

shaharsat committed:
Reordered the creation of the query and the modification of it

shaharsat committed:
Reordered the methods in ElasticLibraryManager to match the order in handlers.go

shaharsat committed:
Replaced redundant error check

shaharsat committed:
Replaced the global initialization of ElasticLibrary

shaharsat committed:
Added a new file that should've been in the previous commit

shaharsat committed:
Moved Redis caching from a global variable to sync.Once

shaharsat committed:
Replaced the global elastic manager instance with a sync.Once

YuvalSteiy commented on cache/redis_request_cacher.go:
```diff
@@ -0,0 +1,51 @@
+	RedisRequestCache *RedisRequestCacher
+)
+
+func NewRedisCache(maxNumber int) *RedisRequestCacher {
```
```suggestion
func NewRedisCacher(maxNumber int) *RedisRequestCacher {
```

YuvalSteiy commented on cache/redis_request_cacher.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+var (
+	Once              sync.Once
```
Did you indent lines?

YuvalSteiy commented on db/elastic_library_manager.go:
```diff
@@ -0,0 +1,159 @@
+
+var (
+	Once              sync.Once
+	ElasticLibraryMan *ElasticLibraryManager
```
```suggestion
	LibraryManager *ElasticLibraryManager
```

YuvalSteiy commented on db/elastic_library_manager.go:
```diff
@@ -0,0 +1,159 @@
+	ElasticLibraryMan *ElasticLibraryManager
+)
+
+func NewElasticLibrary(indexName string) *ElasticLibraryManager {
```
```suggestion
func NewElasticLibrary(indexName string) (*ElasticLibraryManager, error) {
```

YuvalSteiy commented on db/elastic_library_manager.go:
```diff
@@ -0,0 +1,159 @@
+		ElasticLibraryMan = &ElasticLibraryManager{indexName, elasticClient}
+	})
+
+	return ElasticLibraryMan
```
```suggestion
	if ElasticLibraryMan == nil {return errors.New("failed to init ES client...")}
	return ElasticLibraryMan
```

YuvalSteiy commented on db/elastic_library_manager.go:
```diff
@@ -0,0 +1,159 @@
+		elasticClient, err := elastic.NewClient(elastic.SetURL(os.Getenv("ELASTIC_URL")))
+
+		if err != nil {
+			panic(err)
```
Don't panic on error here. Let's pass it back.

shaharsat committed:
Renamed variables and added error handling to ElasticLibrary

shaharsat committed:
Renamed a variable to be more descriptive
